Add support for gRPC transcoding

zZHorizonZz opened the pull request:
Motivation:

This pull request introduces gRPC transcoding capabilities to the gRPC server. The most important changes include the addition of new interfaces for HTTP templates and variable bindings, as well as updates to existing classes to support these new features.

The code for path lookup and checking is based on code from [grpc-transcoding-repo](https://github.com/grpc-ecosystem/grpc-httpjson-transcoding). I think, we shouldn't try to reinvent the wheel when a solution already exists. While there are some changes since the original code is in C++, the logic for matching and extracting HTTP requests remains largely similar. I have also converted several tests from the original repository for these classes.

Regarding the actual transcoding, ~~I'm currently facing two issues. First, by definition, requests can have an empty body while the input gRPC message can still have fields. What I mean is that values can be passed in requests via path or query parameters. However, when I test this, the code gets stuck, my guess is that it's possibly waiting for a body. I need toinvestigate this further but would appreciate any help. The second issue is that in the generator, the code seems unable to read method options despite having them as dependencies~~. But, overall the transcoding seems to be working except for these two issues, though obviously, this pr will need additional work.

Also currently there isn't really support for streaming that would require websockets which is an entirely new thing in my opinion.

The code for the transcoding is reusing the wire format infrastructure by putting length and compression data before the json data, so that we don't need to create entirely new thing for this.

I appreciate any feedback and help with integrating this feature into the codebase.

Closes: https://github.com/eclipse-vertx/vertx-grpc/issues/95

vietj commented:
I need to read more about it, however I think this feature should be implemented in its own module instead of being in grpc common

vietj commented:
I am wondering if we should not make some work on the server to provide an invocation API to decouple the delivery of gRPC invocations from the actual transport.

zZHorizonZz commented:
Are you talking about implementing a middleware layer where you could have daisy chained middlewares with buffers or something, so you could have compression middleware, transcoding middleware, etc.? Or am I thinking of something completely different?

vietj commented:
I'm thinking of a way to community between a gRPC server request/response and an invocation layer using grpc messages, this way anything could invoke a gRPC service by crafting the proper messages, very much like HTTP messages in Netty

zZHorizonZz commented:
I think I might be understanding it, but just maybe....

So if I understand it basically there would be something like this:
```java
public interface MethodRegistry {

<Req, Resp> MethodRegistry callHandler(ServiceMethod<Req, Resp> serviceMethod, Handler<GrpcRequest<Req, Resp>> handler);

}
```

```java
public interface MethodCallHandler<Req, Resp> implements Handler<GrpcRequest<Req, Resp>> {

void handle(GrpcRequest<Req, Resp> grpcRequest);

}
```

So here default implementation if method registry would contain basically logic from server:
```java
if (handler != null) {
      MethodCallHandler<Req, Resp> p = new MethodCallHandler<>(serviceMethod.decoder(), serviceMethod.encoder(), handler);
      methodCallHandlers.compute(serviceMethod.fullMethodName(), (key, prev) -> {
        if (prev == null) {
          prev = new ArrayList<>();
        }
        for (int i = 0; i < prev.size(); i++) {
          MethodCallHandler<?, ?> a = prev.get(i);
          if (a.messageDecoder.format() == serviceMethod.decoder().format() && a.messageEncoder.format() == serviceMethod.encoder().format()) {
            prev.set(i, p);
            return prev;
          }
        }
        prev.add(p);
        return prev;
      });
    } else {
      methodCallHandlers.compute(serviceMethod.fullMethodName(), (key, prev) -> {
        if (prev != null) {
          for (int i = 0; i < prev.size(); i++) {
            MethodCallHandler<?, ?> a = prev.get(i);
            if (a.messageDecoder.format() == serviceMethod.decoder().format() && a.messageEncoder.format() == serviceMethod.encoder().format()) {
              prev.remove(i);
              break;
            }
          }
          if (prev.isEmpty()) {
            prev = null;
          }
        }
        return prev;
      });
    }
    return this;

vietj commented on vertx-grpc-server/src/main/java/io/vertx/grpc/server/impl/GrpcServerImpl.java:
```diff
@@ -206,6 +282,24 @@ public <Req, Resp> GrpcServer callHandler(ServiceMethod<Req, Resp> serviceMethod
+    ServiceTranscodingOptions transcodingOptions) {
+    this.callHandler(serviceMethod, handler);
+
+    if (!options.isGrpcTranscodingEnabled()) {
```
this should throw IllegalStateException

vietj commented:
I'll handle the internal refactoring (decoupling) myself after we merge this PR

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpTemplate.java:
```diff
@@ -0,0 +1,59 @@
+   *
+   * @return The list of segments.
+   */
+  List<String> getSegments();
```
rename `segments`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpTemplate.java:
```diff
@@ -0,0 +1,59 @@
+   *
+   * @return The verb.
+   */
+  String getVerb();
```
rename `verb`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpTemplate.java:
```diff
@@ -0,0 +1,59 @@
+   *
+   * @return The list of variables.
+   */
+  List<HttpTemplateVariable> getVariables();
```
rename `variables`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpTemplateVariable.java:
```diff
@@ -0,0 +1,91 @@
+   *
+   * @return The field path.
+   */
+  List<String> getFieldPath();
```
rename `fieldPath`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpTemplateVariable.java:
```diff
@@ -0,0 +1,91 @@
+   *
+   * @return The starting segment.
+   */
+  int getStartSegment();
```
rename `startSegment`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpTemplateVariable.java:
```diff
@@ -0,0 +1,91 @@
+   *
+   * @return The ending segment.
+   */
+  int getEndSegment();
```
rename `endSegment`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpTemplateVariable.java:
```diff
@@ -0,0 +1,91 @@
+   *
+   * @return {@code true} if the variable represents a wildcard path, {@code false} otherwise.
+   */
+  boolean hasWildcardPath();
```
rename `isWildCardPath`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpVariableBinding.java:
```diff
@@ -0,0 +1,60 @@
+   *
+   * @return A list of field names representing the path in the gRPC message
+   */
+  List<String> getFieldPath();
```
rename `fieldPath`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/HttpVariableBinding.java:
```diff
@@ -0,0 +1,60 @@
+   *
+   * @return The string value to be bound to the specified field path
+   */
+  String getValue();
```
rename `value`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherBuilder.java:
```diff
@@ -0,0 +1,118 @@
+   *
+   * @return the root PathMatcherNode
+   */
+  PathMatcherNode getRoot();
```
rename `root`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherBuilder.java:
```diff
@@ -0,0 +1,118 @@
+   *
+   * @return set of custom verb strings
+   */
+  Set<String> getCustomVerbs();
```
rename `customVerbs`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherBuilder.java:
```diff
@@ -0,0 +1,118 @@
+   *
+   * @return list of PathMatcherMethodData objects
+   */
+  List<PathMatcherMethodData> getMethodData();
```
rename `methodData`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherBuilder.java:
```diff
@@ -0,0 +1,118 @@
+   *
+   * @return the current URL unescaping specification
+   */
+  PercentEncoding.UrlUnescapeSpec getUrlUnescapeSpec();
```
rename `urlUnescapeSpec`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherBuilder.java:
```diff
@@ -0,0 +1,118 @@
+   *
+   * @return current setting for plus sign unescaping in query parameters
+   */
+  boolean getQueryParamUnescapePlus();
```
rename `queryParamUnescapePlus`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherBuilder.java:
```diff
@@ -0,0 +1,118 @@
+   *
+   * @return current setting for matching unregistered custom verbs
+   */
+  boolean getMatchUnregisteredCustomVerb();
```
rename `matchUnregisteredCustomVerb`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherBuilder.java:
```diff
@@ -0,0 +1,118 @@
+   *
+   * @return current setting for failing on duplicate registrations
+   */
+  boolean getFailRegistrationOnDuplicate();
```
rename `failRegisrationOnDuplicate`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherLookupResult.java:
```diff
@@ -0,0 +1,27 @@
+    this.bodyFieldPath = bodyFieldPath;
+  }
+
+  public String getMethod() {
```
rename `method`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherLookupResult.java:
```diff
@@ -0,0 +1,27 @@
+    return method;
+  }
+
+  public List<HttpVariableBinding> getVariableBindings() {
```
rename `variableBindings`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PathMatcherLookupResult.java:
```diff
@@ -0,0 +1,27 @@
+    return variableBindings;
+  }
+
+  public String getBodyFieldPath() {
```
rename `bodyFieldPath`

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/ServiceTranscodingOptions.java:
```diff
@@ -0,0 +1,132 @@
+ *   <li>How to map the HTTP request/response bodies to gRPC messages</li>
+ * </ul>
+ */
+@VertxGen
```
This should be `@DataObject` instead in addition of its components.

We should also make those interfaces plain classes as they are rather data containers and not asynchronous and annotation all of them with `@DataObject`

vietj commented:
Most objects in the transcoding package should be annotated with `@DataObject` and should be plain classes instead of having an interface/implementation

zZHorizonZz commented:
> I'll handle the internal refactoring (decoupling) myself after we merge this PR

I wish I could help, but I think I'm imagining something completely different from what you're thinking, so that will probably be best.

> Most objects in the transcoding package should be annotated with @DataObject and should be plain classes instead of having an interface/implementation

Yeah, I have refactored it so now there are just classes. I did it mostly to fit common package style and also because I had some issues compiling, as I was getting errors about @VertxGen usage, etc., as I've never really used these annotations. But now since it's in its own package, there are no issues.

vietj commented on vertx-grpc-docs/src/main/asciidoc/server.adoc:
```diff
@@ -68,6 +68,12 @@ router.route("/com.mycompany.MyService/*").handler(corsHandler);
 
+==== gRPC Transcoding
+
+The Vert.x gRPC Server supports the gRPC transcoding. The transcoding is disabled by default.
```
link to grpc transcoding

zZHorizonZz replied:
Most of the time when I need to read about transcoding I'm going to google cloud grpc docs as I didn't really find any official transcoding docs. So is this [link](https://cloud.google.com/endpoints/docs/grpc/transcoding) ok are are you have something else in mind.

vietj replied:
My point is that the documentation should do more and explain what transcoding does and when it is useful for users.

zZHorizonZz commented:
I have added documentation. While writing docs for the generator, I have been thinking and this is more or less for discussion if we should add configuration options to enable/disable the generation of wire format and transcoding options. I think in large codebases where these features are not needed, it might make a significant difference in terms of JAR file size.

vietj commented on vertx-grpc-server/src/main/java/io/vertx/grpc/server/impl/GrpcServerImpl.java:
```diff
@@ -106,37 +133,82 @@ public void handle(HttpServerRequest httpRequest) {
+  }
+
+  private <Req, Resp> void handle(MethodCallHandler<Req, Resp> method, HttpServerRequest request, GrpcMethodCall methodCall, PathMatcherLookupResult pathMatcherLookupResult) {
+    if (request.version() == HttpVersion.HTTP_2) {
```
why do we need to not accept HTTP/2 requests?

zZHorizonZz replied:
We don't, I will remove that 👍 I think I initially had this just as check so somebody by accident cannot call this in the future.

vietj commented on vertx-grpc-server/src/main/java/io/vertx/grpc/server/impl/GrpcServerImpl.java:
```diff
@@ -106,37 +133,82 @@ public void handle(HttpServerRequest httpRequest) {
+      return;
+    }
+
+    String contentType = request.getHeader(CONTENT_TYPE);
```
we should send an appropriate HTTP response

vietj commented on vertx-grpc-server/src/main/java/io/vertx/grpc/server/impl/GrpcServerImpl.java:
```diff
@@ -106,37 +133,82 @@ public void handle(HttpServerRequest httpRequest) {
+    }
+
+    MethodTranscodingOptions transcodingOptions = this.transcodingOptions.get(methodCall.fullMethodName());
+    if (transcodingOptions == null) {
```
we should send an appropriate HTTP response

vietj commented on vertx-grpc-server/src/main/java/io/vertx/grpc/server/impl/GrpcServerRequestImpl.java:
```diff
@@ -213,6 +243,61 @@ private boolean notGrpcWebText() {
+      .equals(httpRequest.getHeader(CONTENT_TYPE));
+  }
+
+  private Buffer mutateMessage(Buffer message, List<HttpVariableBinding> bindings) {
```
we should move this code to the transcoding module so it can be tested there independently  of gRPC interactions.

vietj commented:
> I have added documentation. While writing docs for the generator, I have been thinking and this is more or less for discussion if we should add configuration options to enable/disable the generation of wire format and transcoding options. I think in large codebases where these features are not needed, it might make a significant difference in terms of final file size.

we should only perform transcoding generation when the service is annotated with `google.api.http` so the intent is clear

zZHorizonZz commented:
> we should only perform transcoding generation when the service is annotated with google.api.http so the intent is clear

So basically, currently the generator is generating MethodTranscodingOptions for every method and is using default values in these. For example, for the HTTP method it falls back to POST, and for the path it falls back to a path like this:

```java
if (methodContext.transcodingContext.path == null) {
        methodContext.transcodingContext.method = "POST";
        if (serviceContext.packageName == null || serviceContext.packageName.isEmpty()) {
          methodContext.transcodingContext.path = "/" + serviceContext.serviceName + "/" + methodContext.methodName;
        } else {
          methodContext.transcodingContext.path = "/" + serviceContext.packageName + "/" + serviceContext.serviceName + "/" + methodContext.methodName;
        }
      }
```

I think I have an idea on how to generate the code only when there are transcoding annotations available though, so I will try to fix that.

> is grpc.io server generation capable of doing transcoding ? I don't see any the service options metadata available in the generated code

If we are talking about server from for example vertx-grpc-it then there should be generated code on top of the class something like this:
```java
public static final MethodTranscodingOptions SayHello_TRANSCODING = new MethodTranscodingOptions(
    "",
    HttpMethod.valueOf("POST"),
    "/Greeter/SayHello",
    "",
    "",
    List.of(
    ));
 ```
 
 And then some binding methods for transcoding.

vietj commented:
about grpc.io, I am merely wondering whether the transcoding information is available from the grpc-java generator, because I cannot see anything here, so there is no possibility to take a grpc-java service descriptor and obtain the `oogle.api.http` information for honnoring the declared mapping.

vietj commented:
for the vertx generator, we need to honour the `google.api.http` to implement it, that is what users will expect to use

zZHorizonZz commented:
By grpc.io do you mean vertx-grpcio-server module by any chance? I didn't really do anything with that.

vietj commented:
yes I mean vertx-grpcio-server (which uses grpc-java generated code), I am
not asking for anything do be done here (yet) because it seems to me that
the grpc.io generated code does not provide for such options

On Mon, Jan 27, 2025 at 2:30 PM Daniel Fiala ***@***.***>
wrote:

> By grpc.io do you mean vertx-grpcio-server module by any chance? I didn't
> really do anything with that.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse-vertx/vertx-grpc/pull/122#issuecomment-2615769923>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABXDCVBOX5ZSQP4F5BUJS32MYYI7AVCNFSM6AAAAABVUNDUR2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMJVG43DSOJSGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

zZHorizonZz commented:
Ok, so I have fixed the second issue, I had with the generator, so now the generator is generating the transcoding options only for annotated methods. The issue was that I forgot you need to pass the list of options it should be able to read.

zZHorizonZz commented:
Also, I don't know if I should handle this on my own or if you can do that after this is merged, but the files currently don't have license headers, I have included the source of the original files from the gRPC transcoding C++ repo (basically HTTP template classes and path matcher are based on the code from that)  that I rewrote in Java. I'm not sure whether the Apache license of the original repo needs to be defined in the file headers alongside the Eclipse license.

vietj commented:
@zZHorizonZz can you do a clean rebase on latest repo main branch ? that makes it easier to merge and review

vietj commented:
it seems it is doing good progress too :-)

vietj commented on vertx-grpc-it/src/test/proto/helloworld.proto:
```diff
@@ -39,7 +41,9 @@ package helloworld;
   // Sends a greeting
-  rpc SayHello (HelloRequest) returns (HelloReply) {}
+  rpc SayHello (HelloRequest) returns (HelloReply) {
+    option (google.api.http) = { get: "/v1/hello/{name}" };
```
looks great, we need tests for other cases

zZHorizonZz commented:
> @zZHorizonZz can you do a clean rebase on latest repo main branch ? that makes it easier to merge and review

Should I also squash the commits? I don't know what the view on that is in Eclipse projects, but in Quarkus they want that to be done so the git history is more or less clean.

vietj commented:
> > @zZHorizonZz can you do a clean rebase on latest repo main branch ? that makes it easier to merge and review
> 
> Should I also squash the commits? I don't know what the view on that is in Eclipse projects, but in Quarkus they want that to be done so the git history is more or less clean.

we don't mind having commits not squashed if they are properly separated, there is value in the history we might want to preserve sometimes.

we can check when we merge the contribution

vietj commented:
your commit history looks clean, so we might preserve it

vietj commented:
I can see many code reformatted by your IDE which makes the PR harder to review, can you remove those un-necessary changes ?

zZHorizonZz commented:
> I can see many code reformatted by your IDE which makes the PR harder to review, can you remove those un-necessary changes ?

Yeah, sorry about that. IDE is doing its thing where it seems like it's able to adapt to the code style, but at the same time, it's not. I'll fix that.

vietj commented:
@zZHorizonZz I'll make another review soonish, hopefully we can get that in time for 5.0

vietj commented on vertx-grpc-transcoding/src/main/java/module-info.java:
```diff
@@ -0,0 +1,13 @@
+  requires io.netty.transport;
+  requires com.google.protobuf;
+  requires com.google.protobuf.util;
+  requires transitive io.vertx.core;
```
avoid _transitive_ here, for the moment we avoid transitive dependency when declaring module descriptors

vietj commented on vertx-grpc-transcoding/src/main/java/io/vertx/grpc/transcoding/PercentEncoding.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  public static String urlEscapeString(String str) {
+    return Pattern.compile("[^a-zA-Z0-9-_.~]").matcher(str)
```
Cache pattern instance in a static final declaration.

zZHorizonZz replied:
I think we don't need this method at all. It's just that it was in the original file, but I think we might keep it if there's a need for it in the near future. If not, we can remove it later. But I will fix the declaration.

vietj commented:
do we have a test for sub query parameters like here https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L109 ?

zZHorizonZz commented:
> do we have a test for sub query parameters like here https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L109 ?

There is a test in MessageWeaverTest, but not in the actual request testing. I have added a test for that and fixed an issue I found along the way.

vietj commented:
thanks for the contribution I plan to merge it soon

zZHorizonZz commented:
I remembered that I never implemented the HTTP error code mapping, so I have added that. Since there isn't much documentation, it's done on a best effort basis. I followed the mapping from the [.NET implementation of transcoding.](https://github.com/dotnet/aspnetcore/blob/81b289460a9ce5fcbd0d32d80a645f139dffc422/src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs#L122)
